client/client-multi/client-objective-c/Kaa/AbstractHttpChannel.m

```diff
@@ -0,0 +1,293 @@
+/*
+ * Copyright 2014-2015 CyberVision, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#import "AbstractHttpChannel.h"
+#import "IPTransportInfo.h"
+#import "TransportProtocolIdHolder.h"
+#import "KaaLogging.h"
+
+#define TAG @"AbstractHttpChannel >>>"
+
+#define UNATHORIZED_HTTP_STATUS 401
+
+@interface AbstractHttpChannel ()
+
+@property (nonatomic,strong) IPTransportInfo *currentServer;
+@property (nonatomic,weak) AbstractKaaClient *kaaClient;
+@property (nonatomic,strong) id<KaaClientState> kaaState;
+@property (nonatomic,strong) id<FailoverManager> failoverMgr;
+
+@property (nonatomic,strong) volatile NSOperationQueue *executor;
+
+@property (nonatomic) volatile BOOL lastConnectionFailed;
+@property (nonatomic) volatile BOOL isPaused;
+
+@property (nonatomic,strong) AbstractHttpClient *kaaHttpClient;
+@property (nonatomic,strong) id<KaaDataDemultiplexer> chDemultiplexer;
+@property (nonatomic,strong) id<KaaDataMultiplexer> chMultiplexer;
+
+@end
+
+@implementation AbstractHttpChannel
+
+- (instancetype)initWithClient:(AbstractKaaClient *)client state:(id<KaaClientState>)state
+               failoverManager:(id<FailoverManager>)manager {
+    self = [super init];
+    if (self) {
+        self.kaaClient = client;
+        self.kaaState = state;
+        self.failoverMgr = manager;
+    }
+    return self;
+}
+
+- (TransportProtocolId *)getTransportProtocolId {
+    return [TransportProtocolIdHolder HTTPTransportID];
+}
+
+- (void)sync:(TransportType)type {
+    @synchronized(self) {
+        [self syncTransportTypes:[NSSet setWithObject:[NSNumber numberWithInt:type]]];
+    }
+}
+
+- (void)syncTransportTypes:(NSSet *)types {
+    @synchronized(self) {
+        if (self.isShutdown) {
```

Comment:
Yes, this may require same refactoring like was done to DefaultOperationTcpChannel at some point of time (i.e. replace boolean with state enum)